packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts

```diff
@@ -584,3 +610,84 @@ class ImportedServerlessCluster extends ServerlessClusterBase implements IServer
     return this._clusterReadEndpoint;
   }
 }
+
+/**
+ * Properties for ``ServerlessClusterFromSnapshot``
+ */
+export interface ServerlessClusterFromSnapshotProps extends ServerlessClusterBaseProps {
+  /**
+   * The identifier for the DB instance snapshot or DB cluster snapshot to restore from.
+   * You can use either the name or the Amazon Resource Name (ARN) to specify a DB cluster snapshot.
+   * However, you can use only the ARN to specify a DB instance snapshot.
+   */
+  readonly snapshotIdentifier: string;
+
+  /**
+   * Master user credentials.
+   *
+   * Note - It is not possible to change the master username for a snapshot;
+   * however, it is possible to provide (or generate) a new password.
+   *
+   * @default - The existing username and password from the snapshot will be used.
+   */
+  readonly credentials?: SnapshotCredentials;
+}
+
+/**
+ * A Aurora Serverless Cluster restored from a snapshot.
+ *
+ * @resource AWS::RDS::DBCluster
+ */
+export class ServerlessClusterFromSnapshot extends ServerlessClusterNew {
+  public readonly clusterIdentifier: string;
+  public readonly clusterEndpoint: Endpoint;
+  public readonly clusterReadEndpoint: Endpoint;
+  public readonly connections: ec2.Connections;
+  protected enableDataApi?: boolean;
+  public readonly secret?: secretsmanager.ISecret;
+
+  constructor(scope: Construct, id: string, props: ServerlessClusterFromSnapshotProps) {
+    super(scope, id, props);
+
+    this.enableDataApi = props.enableDataApi;
+
+    let credentials = props.credentials;
+    let secret = credentials?.secret;
+    if (!secret && credentials?.generatePassword) {
+      if (!credentials.username) {
+        throw new Error('`credentials` `username` must be specified when `generatePassword` is set to true');
+      }
+
+      secret = new DatabaseSecret(this, 'Secret', {
+        username: credentials.username,
+        encryptionKey: credentials.encryptionKey,
+        excludeCharacters: credentials.excludeCharacters,
+        replaceOnPasswordCriteriaChanges: credentials.replaceOnPasswordCriteriaChanges,
+        replicaRegions: credentials.replicaRegions,
+      });
+    }
+
+    const cluster = new CfnDBCluster(this, 'Resource', {
+      ...this.newCfnProps,
+      snapshotIdentifier: props.snapshotIdentifier,
+      masterUserPassword: secret?.secretValueFromJson('password')?.toString() ?? credentials?.password?.toString(),
+    });
+
+    this.clusterIdentifier = cluster.ref;
+
+    // create a number token that represents the port of the cluster
+    const portAttribute = Token.asNumber(cluster.attrEndpointPort);
+    this.clusterEndpoint = new Endpoint(cluster.attrEndpointAddress, portAttribute);
+    this.clusterReadEndpoint = new Endpoint(cluster.attrReadEndpointAddress, portAttribute);
+    this.connections = new ec2.Connections({
+      securityGroups: this.securityGroups,
+      defaultPort: ec2.Port.tcp(this.clusterEndpoint.port),
+    });
```

Comment:
Can we move this up to `ServerlessClusterNew`? I believe this is duplicated between `ServerlessClusterFromSnapshot` and `ServerlessCluster`.

Reply:
Done.